src/plugins/ui_feature/find_and_replace.ts

```diff
@@ -138,37 +154,119 @@ export class FindAndReplacePlugin extends UIPlugin {
     }
     this.currentSearchRegex = RegExp(searchValue, flags);
   }
-  /**
-   * Find matches using the current regex
-   */
+
+  // Find matches using the current regex
   private findMatches() {
-    const sheetId = this.getters.getActiveSheetId();
-    const cells = this.getters.getCells(sheetId);
     const matches: SearchMatch[] = [];
 
-    if (this.toSearch) {
-      for (const cell of Object.values(cells)) {
-        const { col, row } = this.getters.getCellPosition(cell.id);
-        if (
-          cell &&
-          this.currentSearchRegex &&
-          this.currentSearchRegex.test(this.getSearchableString({ sheetId, col, row }))
-        ) {
-          const match: SearchMatch = { col, row, selected: false };
-          matches.push(match);
+    if (!this.toSearch) {
+      return matches;
+    }
+
+    switch (this.searchOptions.searchScope) {
+      case "allSheets":
+        const sheets = this.getters.getSheetIds();
+        for (const [sheetIndex, sheet] of sheets.entries()) {
+          this.findMatchesInSheet(sheet, sheetIndex, matches);
         }
-      }
+        break;
+      case "thisSheet":
+        const sheetId = this.getters.getActiveSheetId();
+        const sheetIndex = this.getters.getSheetIds().indexOf(sheetId);
+        this.findMatchesInSheet(sheetId, sheetIndex, matches);
+        break;
+      case "specificRange":
+        const specificRange = this.searchOptions.specificRange;
+        if (specificRange) {
+          let rangeSheet, rangeCells;
+
+          if (specificRange.includes("!")) {
+            [rangeSheet, rangeCells] = specificRange.split("!");
+          } else {
+            rangeSheet = this.getters.getSheetName(this.getters.getActiveSheetId());
+            rangeCells = specificRange;
+          }
+
+          const [startCell, endCell] = rangeCells.split(":");
+
+          const finalEndCell = endCell || startCell;
+          const specificRangeWithEndCell = `${startCell}:${finalEndCell}`;
```

Comment:
We obviously have helpers to do this, there's no way we'd want to parse a string into a zone by hand... You even used one (even if it was the wrong one) in `findMatchesInSheet`, why would you do it by hand here ??